File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        int horizontalLevel = 0;
        ArrayList<Integer> ans = new ArrayList<Integer>();
        reversePreOrder(root,horizontalLevel,ans);

        return ans;
    }
    public void reversePreOrder(TreeNode node, int depth, ArrayList<Integer> ans){
        if(node == null){
            return;
        }
        if(depth == ans.size()){
            ans.add(node.val);
        }

        reversePreOrder(node.right,depth+1,ans);
        reversePreOrder(node.left,depth+1,ans);
    }
}

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if (!root) {
            return {};
        }
        deque<TreeNode*> q;
        q.push_back(root);
        vector<int> ans;
        rightView(q, ans);
        return ans;
    }
    void rightView(deque<TreeNode*> q, vector<int>& ans) {
        while (!q.empty()) {
            int n = q.size();
            TreeNode* back = q.back();
            int backVal = back->val;
            ans.push_back(backVal);
            for (int i = 0; i < n; i++) {
                TreeNode* front = q.front();
                q.pop_front();
                if (front->left) {
                    q.push_back(front->left);
                }
                if (front->right) {
                    q.push_back(front->right);
                }
            }
        }
    }
};